### ISSUE TITLE
Insert JSON object into distributed table

### ISSUE THREAD

**Issue by PeerXu** [ClickHouse/ClickHouse#37584]
**Describe what's wrong**

Hi, everyone.

When I try to insert JSON object into distributed table, it throw an expection:

```
ColumnObject must be converted to ColumnTuple before use.
```

**Does it reproduce on recent release?**

Yes, `v22.5.1.2079-stable`

**How to reproduce**

* Which ClickHouse server version to use

v22.5.1.2079-stable

* Which interface to use, if matters

clickhouse-client

* Non-default settings, if any

set allow_experimental_object_type=1;

* `CREATE TABLE` statements for all tables involved

```
CREATE TABLE t1_local
ON CLUSTER '{cluster}'
(
    `o` Object('json')
)
ENGINE = ReplicatedMergeTree('/clickhouse/chi/{installation}/{cluster}/tables/{shard}/default/t1_local', '{replica}')
ORDER BY tuple()

CREATE TABLE t1_distributed
ON CLUSTER '{cluster}'
AS t1_local
ENGINE = Distributed('{cluster}', 'default', 't1_local', rand()) 
```

* Queries to run that lead to unexpected result

```
$ clickhouse-client -q "insert into t1_distributed values ('{\"x\":1}');"
Received exception from server (version 22.5.1):
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: ColumnObject must be converted to ColumnTuple before use. (LOGICAL_ERROR)
(query: insert into t1_distributed values ('{"x":1}');)

```

**Expected behavior**

JSON object should be insert into `t1_local` through with `t1_distributed`.

**Error message and/or stacktrace**

```
2022.05.27 03:02:08.337890 [ 98 ] {} <Debug> DiskLocal: Reserving 1.00 MiB on disk `default`, having unreserved 1023.94 TiB.
2022.05.27 03:02:09.814025 [ 262 ] {} <Debug> TCP-Session: 61f01724-438b-460c-a378-0aaf0edcac82 Creating query context from session context, user_id: ce3b5f2c-37af-608e-6c19-d6deb1384975, parent context user: demo
2022.05.27 03:02:09.814313 [ 262 ] {da115fcf-06cb-4be3-8def-ba7731c3bee2} <Debug> executeQuery: (from [::ffff:127.0.0.1]:40610, user: demo) insert into t1_distributed values 
2022.05.27 03:02:09.816048 [ 262 ] {da115fcf-06cb-4be3-8def-ba7731c3bee2} <Error> executeQuery: Code: 49. DB::Exception: ColumnObject must be converted to ColumnTuple before use. (LOGICAL_ERROR) (version 22.5.1.2079 (official build)) (from [::ffff:127.0.0.1]:40610) (in query: insert into t1_distributed values ), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xb4903fa in /usr/bin/clickhouse
1. DB::ColumnObject::throwMustBeConcrete() @ 0x161e9194 in /usr/bin/clickhouse
2. ? @ 0x161e7b26 in /usr/bin/clickhouse
3. DB::DistributedSink::splitBlock(DB::Block const&) @ 0x1661dfb7 in /usr/bin/clickhouse
4. DB::DistributedSink::writeSplitAsync(DB::Block const&) @ 0x1661b7c0 in /usr/bin/clickhouse
5. DB::DistributedSink::consume(DB::Chunk) @ 0x16619b57 in /usr/bin/clickhouse
6. DB::SinkToStorage::onConsume(DB::Chunk) @ 0x16f4471d in /usr/bin/clickhouse
7. ? @ 0x16edb3d7 in /usr/bin/clickhouse
8. ? @ 0x16edb189 in /usr/bin/clickhouse
9. DB::ExceptionKeepingTransform::work() @ 0x16edaa3f in /usr/bin/clickhouse
10. DB::ExecutionThreadContext::executeTask() @ 0x16d18ee8 in /usr/bin/clickhouse
11. DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x16d0cbfe in /usr/bin/clickhouse
12. DB::PipelineExecutor::executeStep(std::__1::atomic<bool>*) @ 0x16d0c400 in /usr/bin/clickhouse
13. DB::TCPHandler::processInsertQuery() @ 0x16cbc25c in /usr/bin/clickhouse
14. DB::TCPHandler::runImpl() @ 0x16cb53bf in /usr/bin/clickhouse
15. DB::TCPHandler::run() @ 0x16cc55d9 in /usr/bin/clickhouse
16. Poco::Net::TCPServerConnection::start() @ 0x1b0bbc2f in /usr/bin/clickhouse
17. Poco::Net::TCPServerDispatcher::run() @ 0x1b0be081 in /usr/bin/clickhouse
18. Poco::PooledThread::run() @ 0x1b284169 in /usr/bin/clickhouse
19. Poco::ThreadImpl::runnableEntry(void*) @ 0x1b2814c0 in /usr/bin/clickhouse
20. ? @ 0x7f507b134609 in ?
21. __clone @ 0x7f507b059163 in ?

2022.05.27 03:02:09.816165 [ 262 ] {da115fcf-06cb-4be3-8def-ba7731c3bee2} <Error> TCPHandler: Code: 49. DB::Exception: ColumnObject must be converted to ColumnTuple before use. (LOGICAL_ERROR), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xb4903fa in /usr/bin/clickhouse
1. DB::ColumnObject::throwMustBeConcrete() @ 0x161e9194 in /usr/bin/clickhouse
2. ? @ 0x161e7b26 in /usr/bin/clickhouse
3. DB::DistributedSink::splitBlock(DB::Block const&) @ 0x1661dfb7 in /usr/bin/clickhouse
4. DB::DistributedSink::writeSplitAsync(DB::Block const&) @ 0x1661b7c0 in /usr/bin/clickhouse
5. DB::DistributedSink::consume(DB::Chunk) @ 0x16619b57 in /usr/bin/clickhouse
6. DB::SinkToStorage::onConsume(DB::Chunk) @ 0x16f4471d in /usr/bin/clickhouse
7. ? @ 0x16edb3d7 in /usr/bin/clickhouse
8. ? @ 0x16edb189 in /usr/bin/clickhouse
9. DB::ExceptionKeepingTransform::work() @ 0x16edaa3f in /usr/bin/clickhouse
10. DB::ExecutionThreadContext::executeTask() @ 0x16d18ee8 in /usr/bin/clickhouse
11. DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x16d0cbfe in /usr/bin/clickhouse
12. DB::PipelineExecutor::executeStep(std::__1::atomic<bool>*) @ 0x16d0c400 in /usr/bin/clickhouse
13. DB::TCPHandler::processInsertQuery() @ 0x16cbc25c in /usr/bin/clickhouse
14. DB::TCPHandler::runImpl() @ 0x16cb53bf in /usr/bin/clickhouse
15. DB::TCPHandler::run() @ 0x16cc55d9 in /usr/bin/clickhouse
16. Poco::Net::TCPServerConnection::start() @ 0x1b0bbc2f in /usr/bin/clickhouse
17. Poco::Net::TCPServerDispatcher::run() @ 0x1b0be081 in /usr/bin/clickhouse
18. Poco::PooledThread::run() @ 0x1b284169 in /usr/bin/clickhouse
19. Poco::ThreadImpl::runnableEntry(void*) @ 0x1b2814c0 in /usr/bin/clickhouse
20. ? @ 0x7f507b134609 in ?
21. __clone @ 0x7f507b059163 in ?
```

**Comment by CurtizJ** [comment]
Insert into Distributed tables is not implemented for type JSON yet.

**Comment by fizerkhan** [comment]
@CurtizJ Do you have any ETA for this?

**Comment by yandooo** [comment]
The important ~experimental~ feature that doesn't work for distributed tables. It turned out no information in official docs about limitations.

**Comment by adam-mrozik** [comment]
Bumping, would love to use both together as well

**Comment by sundarv85** [comment]
Any timeline on when this would be available..

**Comment by alvinhom** [comment]
I appreciate a timeline on this as well, as I need to determined whether to go with JSON based schema or use Arrays.  JSON is much more efficient, but need to support distributed.

**Comment by sundarv85** [comment]
Just pinging back to check the roadmap as to when this would become available in a distributed format..

**Comment by dilyevsky** [comment]
It appears same issue affects Live Views created on a table with JSON objects. When trying to WATCH such view:

```
Code: 48. DB::Exception: Received from localhost:9000. DB::Exception: ColumnObject must be converted to ColumnTuple before use. (NOT_IMPLEMENTED)
```

**Comment by shengjiangfeng** [comment]
Now is 2025，any update on this ?

**Comment by alexey-milovidov** [comment]
@shengjiangfeng, this problem is fixed in ClickHouse version 24.12.
